optaplanner-core/src/main/java/org/optaplanner/core/impl/AbstractFromConfigFactory.java

```diff
@@ -0,0 +1,115 @@
+
+    protected GenuineVariableDescriptor deduceVariableDescriptor(EntityDescriptor entityDescriptor, String variableName) {
+        GenuineVariableDescriptor variableDescriptor;
+        variableDescriptor = entityDescriptor.getGenuineVariableDescriptor(Objects.requireNonNull(variableName));
+        if (variableDescriptor == null) {
+            throw new IllegalArgumentException("The config (" + config
+                    + ") has a variableName (" + variableName
+                    + ") which is not a valid planning variable on entityClass ("
+                    + entityDescriptor.getEntityClass() + ").\n"
+                    + entityDescriptor.buildInvalidVariableNameExceptionMessage(variableName));
+        }
+        return variableDescriptor;
+    }
+
+    protected GenuineVariableDescriptor deduceVariableDescriptor(EntityDescriptor entityDescriptor) {
+        Collection<GenuineVariableDescriptor> variableDescriptors = entityDescriptor.getGenuineVariableDescriptors();
+        if (variableDescriptors.size() != 1) {
+            throw new IllegalArgumentException("The config (" + config
+                    + ") has no configured variableName for entityClass (" + entityDescriptor.getEntityClass()
+                    + ") and because there are multiple variableNames ("
+                    + entityDescriptor.getGenuineVariableNameSet()
+                    + "), it cannot be deduced automatically.");
+        }
+        return variableDescriptors.iterator().next();
+    }
+
+    protected List<GenuineVariableDescriptor> deduceVariableDescriptorList(EntityDescriptor entityDescriptor,
+            List<String> variableNameIncludeList) {
+        Objects.requireNonNull(entityDescriptor);
+        List<GenuineVariableDescriptor> variableDescriptorList = entityDescriptor.getGenuineVariableDescriptorList();
+        if (variableNameIncludeList == null) {
+            return variableDescriptorList;
+        }
+        List<GenuineVariableDescriptor> resolvedVariableDescriptorList =
+                new ArrayList<>(variableDescriptorList.size());
+        for (String variableNameInclude : variableNameIncludeList) {
+            boolean found = false;
+            for (GenuineVariableDescriptor variableDescriptor : variableDescriptorList) {
+                if (variableDescriptor.getVariableName().equals(variableNameInclude)) {
+                    resolvedVariableDescriptorList.add(variableDescriptor);
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                throw new IllegalArgumentException("The config (" + this
+                        + ") has a variableNameInclude (" + variableNameInclude
+                        + ") which does not exist in the entity (" + entityDescriptor.getEntityClass()
+                        + ")'s variableDescriptorList (" + variableDescriptorList + ").");
+            }
+        }
+        return resolvedVariableDescriptorList;
```

Comment:
This could be a nice stream.

    return variableNameIncludeList.stream()
        .map(n -> variableDescriptorList.stream()
            .filter(d -> d.gerVariableName().equals(n))
            .findFirst()
            .orElseThrow(...))
        .collect(Collectors.toList());

In my opinion, this reads better as you don't need to mentally track the state of `found`.

Reply:
I didn't touch the existing code if I wasn't changing it. That said, if there is an opportunity to improve it, I will take a look at this one and other suggestions you made.